### PR TITLE
Sourcemap docs

### PR DESCRIPTION
**Pull request by DeMoorJasper**
Some initial docs for the sourcemaps section, still need to double check my text to make sure I haven't made any great mistakes

### CODE REVIEW

File: src/plugin-system/source-maps.md [hunk 1]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
   order: 14
 ---
 
-TODO
+Parcel utilizes the package `@parcel/source-maps` for all it's source-maps, this ensures performance and reliability when manipulating sourcemaps across plugins and Parcel's core. This library has been written from the ground up in C++ with manipulation and concatenation of these sourcemaps in mind and gave us a 20x performance improvement over our old solution consisting of Mozilla's Source-Map library and some internal utilities. This improvement in performance is mainly due to the optimizations in the data structures and way we cache the sourcemaps.
```
Comment (mischnic) [inline src/plugin-system/source-maps.md]:
```suggestion
Parcel utilizes the package `@parcel/source-maps` for processing all source maps to ensure performance and reliability when manipulating source maps across plugins and Parcel's core. This library has been written from the ground up in C++ with both source map manipulation and concatenation in mind and gave us a 20x performance improvement over our old solution using Mozilla's [`source-map`](https://github.com/mozilla/source-map) library and some internal utilities. This improvement in performance is mainly due to optimizations in the data structures and the way in which we cache source maps.
```

File: src/plugin-system/source-maps.md [hunk 2]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
...
+
+## How to use the library
+
+To use the library you start of by creating an instance of the exported `SourceMap` class, once you've done that you can call various functions to add and edit mappings.
```
Comment (mischnic) [inline src/plugin-system/source-maps.md]:
```suggestion
To use the library, you start off by creating an instance of the exported `SourceMap` class, on which you can call various functions to add and edit source mappings.
```

File: src/plugin-system/source-maps.md [hunk 3]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
...
+
+Below is an example with all the ways you can add mappings to a SourceMap instance:
```
Comment (mischnic) [inline src/plugin-system/source-maps.md]:
```suggestion
Below is an example covering all ways of adding mappings to a `SourceMap` instance:
```

File: src/plugin-system/source-maps.md [hunk 4]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
...
+
+```js
+import SourceMap from "@parcel/source-map";
+
+let sourcemap = new SourceMap();
+
+// Each function that adds mappings, has optional offset arguments
+// these can be used to offset the generated mappings by a certain amount
```
Comment (mischnic) [inline src/plugin-system/source-maps.md]:
```suggestion
// Each function that adds mappings has optional offset arguments.
// These can be used to offset the generated mappings by a certain amount.
```

File: src/plugin-system/source-maps.md [hunk 5]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
...
+
+// Sourcemaps can be saved as buffers (flatbuffers), this is what we use for caching in Parcel.
+// You can instantiate a SourceMap with these buffer values using the `addBufferMappings` function
+let originalMapBuffer = new Buffer();
+sourcemap.addBufferMappings(originalMapBuffer, lineOffset, columnOffset);
+```
+
+## Transformations/Manipulations
+
+If your plugin does any code manipulations it should ensure that it creates correct mappings to the original source-code to guarantee that we still end up creating an accurate source-map at the end of the bundling process. You are expected to return a `SourceMap` instance at the end of a transform in a [Transformer](/plugin-system/transformer/) plugin. We also provide the source-map from the previous transform to ensure you map to the original source code and not the output of the previous transform.
```
Comment (mischnic) [inline src/plugin-system/source-maps.md]:
```suggestion
If your plugin does any code manipulations, you should ensure that it creates correct mappings to the original source code to guarantee that we still end up creating an accurate source map at the end of the bundling process. You are expected to return a `SourceMap` instance at the end of a transform in a [Transformer](/plugin-system/transformer/) plugin. We also provide the source 
 map from the previous transform to ensure you map to the original source code and not just the output of the previous transform.
```

File: src/plugin-system/source-maps.md [hunk 6]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
...
+      }
+    });
+
+    // Return the contents and map so Parcel Core can save these to disk or get post-processed by optimizers
+    return { contents, map };
+  },
+});
+```
+
+## PostProcessing sourcemaps in Optimizers
```
Comment (mischnic) [inline src/plugin-system/source-maps.md]:
```suggestion
## Postprocessing source maps in optimizers
```

File: src/plugin-system/source-maps.md [hunk 7]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
...
+
+Using sourcemaps in optimizers is identical to how you use it in Transformers as you get one file as input and are expected to return that same file as output but optimized.
```
Comment (mischnic) [inline src/plugin-system/source-maps.md]:
```suggestion
Using sourcemaps in optimizers is identical to how you use it in transformers as you get one file as input and are expected to return that same file as output but optimized.
```

File: src/plugin-system/source-maps.md [hunk 8]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
...
+```js
+// The contents and map are passed separately
+async optimize({ bundle, contents, map }) {
+  return { contents, map }
+}
+```
+
+## Diagnosing issues
+
+If you encounter incorrect mappings and want to debug these mappings we have built tools that can help you diagnose these issues, this includes a specific reporter (`@parcel/reporter-sourcemap-visualiser`) to create a `sourcemap-info.json` file with all the necessary information to visualize all the mappings and source content.
```
Comment (mischnic) [inline src/plugin-system/source-maps.md]:
```suggestion
If you encounter incorrect mappings and want to debug these mappings we have built tools that can help you diagnose these issues. By running a specific reporter (`@parcel/reporter-sourcemap-visualiser`), Parcel create a `sourcemap-info.json` file with all the necessary information to visualize all the mappings and source content.
```

File: src/plugin-system/source-maps.md [hunk 9]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
...
+
+For it to work you can add a custom `.parcelrc`:
```
Comment (mischnic) [inline src/plugin-system/source-maps.md]:
```suggestion
To enable it, add a custom `.parcelrc`:
```

File: src/plugin-system/source-maps.md [hunk 10]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
...
+If your plugin does any code manipulations, you should ensure that it creates correct mappings to the original source code to guarantee that we still end up creating an accurate source map at the end of the bundling process. You are expected to return a `SourceMap` instance at the end of a transform in a [Transformer](/plugin-system/transformer/) plugin. We also provide the source map from the previous transform to ensure you map to the original source code and not just the output of the previous transform.
+
+The `asset` value that gets passed in the `parse`, `transform` and `generate` functions of a transformer plugin contains a function called `getMap()` and `getMapBuffer()`, these functions can be used to get a SourceMap instance (`getMap()`) and the cached SourceMap Buffer (`getMapBuffer()`).
+
+You are free to manipulate the sourcemap at any of these steps in the transformer as long as you ensure the sourcemap that gets returned in `generate` maps to the original sourcefile correctly.
+
+Below is an example on how to manipulate sourcemaps in a transformer plugin:
+
+```js
+import { Transform } from "@parcel/plugin";
```
Comment (mischnic) [inline src/plugin-system/source-maps.md]:
```suggestion
import { Transformer } from "@parcel/plugin";
```

File: src/plugin-system/source-maps.md [hunk 11]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
...
+import SourceMap from "@parcel/source-map";
+
+export default new Transform({
```
Comment (mischnic) [inline src/plugin-system/source-maps.md]:
```suggestion
export default new Transformer({
```

File: src/plugin-system/source-maps.md [hunk 12]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
...
+      code: compilationResult.code,
+      map,
+    };
+  },
+});
+```
+
+## Concatenating sourcemaps in Packagers
+
+If you're writing a custom packager, it's your responsibility to concatenate the sourcemaps of all the assets while packaging the assets. This is done by creating a new sourcemap instance using `new SourceMap()` and adding new mappings to it using the `addBufferMappings(buffer, lineOffset, columnOffset)` function. The lineOffset should be equal to the line index at which the asset output starts.
```
Comment (mischnic) [inline src/plugin-system/source-maps.md]:
```suggestion
If you're writing a custom packager, it's your responsibility to concatenate the sourcemaps of all the assets while packaging the assets. This is done by creating a new `SourceMap` instance and adding new mappings to it using the `addBufferMappings(buffer, lineOffset, columnOffset)` function. `lineOffset` should be equal to the line index at which the asset output starts.
```

File: src/plugin-system/source-maps.md [hunk 13]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
...
+        i++;
+      }
+    });
+
+    // Return the contents and map so Parcel Core can save these to disk or get post-processed by optimizers
+    return { contents, map };
+  },
+});
+```
+
```
Comment (mischnic) [inline src/plugin-system/source-maps.md:227]: Maybe a short paragraph that for more complicated situations where packaging happens on an AST level, it works similar to the transformer?

File: src/plugin-system/source-maps.md [hunk 14]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
...
+## Postprocessing source maps in optimizers
+
+Using sourcemaps in optimizers is identical to how you use it in transformers as you get one file as input and are expected to return that same file as output but optimized.
```
Comment (mischnic) [inline src/plugin-system/source-maps.md]:
```suggestion
Using source maps in optimizers is identical to how you use it in transformers as you get one file as input and are expected to return that same file as output but optimized.
```

File: src/plugin-system/source-maps.md [hunk 15]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
...
+
+The only difference with optimizers is that the map is not provided as part of an asset but rather as a separate parameter/option as you can see in the code snippet below. As always the map is an instance of the Parcel SourceMap class.
```
Comment (mischnic) [inline src/plugin-system/source-maps.md]:
```suggestion
The only difference with optimizers is that the map is not provided as part of an asset but rather as a separate parameter/option as you can see in the code snippet below. As always, the map is an instance of the `SourceMap` class.
```

File: src/plugin-system/source-maps.md [hunk 16]
```diff
@@ -6,7 +6,252 @@ eleventyNavigation:
...
+To enable it, add a custom `.parcelrc`:
+
+```json
+{
+  "extends": "@parcel/config-default",
+  "reporters": ["...", "@parcel/reporter-sourcemap-visualiser"]
+}
+```
+
+After the reporter has created the `sourcemap-info.json` file you can upload it to the [sourcemap visualizer](https://sourcemap-visualiser.now.sh/)
```
Comment (mischnic) [inline src/plugin-system/source-maps.md]:
```suggestion
After the reporter has created the `sourcemap-info.json` file, you can upload it to the [sourcemap visualizer](https://sourcemap-visualiser.now.sh/)
```

File: src/plugin-system/source-maps.md (hunk 10 above)
Comment (DeMoorJasper) [reply src/plugin-system/source-maps.md]: this was copy-paste from the transformer docs, will update those as well in that case. Forgot I renamed it

File: src/plugin-system/source-maps.md (hunk 13 above)
Comment (DeMoorJasper) [reply src/plugin-system/source-maps.md:227]: Sure will add that in

File: src/plugin-system/source-maps.md [hunk 17]
```diff
@@ -6,7 +6,282 @@ eleventyNavigation:
...
+      }
+    });
+
+    // Return the contents and map so Parcel Core can save these to disk or get post-processed by optimizers
+    return { contents, map };
+  },
+});
+```
+
+### Concatenating ASTs
```
Comment (DeMoorJasper) [inline src/plugin-system/source-maps.md:228]: Added this section as requested, didn't go very in depth as this is probably something for the Packager docs, mainly just touched that it works similar to the Transformer and the risk of losing mappings when doing lots of modifications. The mapping side of things is very simple though, it's just compile turn into SourceMap object and return

Comment (mischnic) [reply src/plugin-system/source-maps.md:228]: 👍  I just wanted it to mention that this is also a common approach

**Review by mischnic — APPROVED** [review]

**Pull request merged**